### PR TITLE
refactor: pkg_resources -> importlib.metadata

### PR DESCRIPTION
**Pull request by cwegener**
pkg_resources is in the process of being deprecated in favor of
importlib.metadata and importlib.resources

importlib.metadata has been available in the standard library since 3.8

And as of 3.10 the importlib.metadata API is no longer provisional

In order to support Python version older than 3.10, the
`importlib_metadata` backport is used instead of the actual
`importlib.metadata` API.

### CODE REVIEW

**Comment by cwegener** [comment]
> Thanks @cwegener for the PR. Note you'll also have to re-compile the Python dependencies which ensures we have a frozen set of Python packages (and sub-packages) for testing purposes. Please refer to the CONTRIBUTING.md file with regards to how to do this.

Hi John. I tried, but pip-compile-multi seems to just blissfully ignore the install_dependencies from setup.py

Do you have any hints on how to make pip-compile-multi read the install_dependencies collection from setup.py?

**Comment by john-bodley** [comment]
@cwegener I gather you mean `install_requirements` rather than `install_dependencies`? Note that `pip-compile-multi` should adhere to the requirements. Would you mind sharing the command(s) you ran?

**Comment by cwegener** [comment]
> @cwegener I gather you mean `install_requirements` rather than `install_dependencies`? Note that `pip-compile-multi` should adhere to the requirements. Would you mind sharing the command(s) you ran?

Sorry, yes. Typo. `install_requires` is what I'm referring to.
All good. It turns out that the result of `pip-compile-multi -P importlib-metadata` is simply a noop, because there is no change to the version of `importlib-metadata` that is already present.
And `pip-compile-multi` just doesn't need to know that `importlib-medata` is now a top-level require instead of a transitive one.

**Comment by cwegener** [comment]
I did just force push the 'noop' result from pip-compile-multi, which is basically just adding the comment into the requirements/base.txt for good measure.

File: requirements/base.txt [hunk 1]
```diff
@@ -137,7 +137,9 @@ humanize==3.11.0
 idna==3.2
     # via email-validator
 importlib-metadata==6.6.0
-    # via flask
+    # via
```
Comment (john-bodley) [inline requirements/base.txt:140]: Yes!

File: requirements/base.txt [hunk 2]
```diff
@@ -137,7 +137,9 @@ humanize==3.11.0
...
+    #   apache-superset
```
Comment (john-bodley) [inline requirements/base.txt:141]: Yes!

**Comment by cwegener** [comment]
On a different note, once I'm done with this `pkg_resources` migration (I'll publish the other part which is the migration to importlib.resources shortly) and setuptools should then no longer be a dependency,I hope to convince the superset contributors to migrate from setuptools+pip-compile-multi to hatch+pdm ... :smile:

**Comment by cwegener** [comment]
@john-bodley Sorry about the missing import sort. I just pushed the fix for that. Can you run the GH actions again.

**Comment by john-bodley** [comment]
@cwegener with regards to, 

> I hope to convince the superset contributors to migrate from setuptools+pip-compile-multi to hatch+pdm ...

it might make sense to first create an issue (or possibly a SIP dependency on the complexity) regarding your intentions as it's probably worth discussing the pros/cons especially given the slew of Python package managers floating around.

**Review by john-bodley — APPROVED** [review]

**Comment by john-bodley** [comment]
@cwegener after rereading your PR description you mention that,

> importlib.metadata has been available in the standard library since 3.8

and given we only support Python 3.9+ I was wondering why we need to use the backport.

**Comment by cwegener** [comment]
> and given we only support Python 3.9+ I was wondering why we need to use the backport.

The backport is needed because the API of `importlib.metadata` is actually different in 3.9. The API has only been stable in 3.10.

Specifically, the `entry_points()` signature in 3.9 was still different, it didn't have the `groups` parameter in the 3.9 standard library. So the stable API is only available in the backport. (Found that out when I developed this change in 3.10 accidentally and then ran unit tests in 3.9)

**Comment by cwegener** [comment]
> Specifically, the `entry_points()` signature in 3.9 was still different, it didn't have the `groups` parameter in the 3.9 standard library. So the stable API is only available in the backport. (Found that out when I developed this change in 3.10 accidentally and then ran unit tests in 3.9)

The best I can tell, the 'select' feature wasn't there in the `entry_points()`  

I did just see though that there also is a more specific backport of JUST the select feature for entry_points() :grin:

https://github.com/jaraco/backports.entry_points_selectable

**Pull request merged**